### PR TITLE
Yell at the user for combining ICs and restart

### LINKED ISSUE #21423: Checkpoint: no log or warning on conflict between restart and initial conditions
## Bug Description
When using the checkpoint restart system, the initial conditions are likely not to be used. There is nothing telling the user that they are used anyway

## Steps to Reproduce
Do a restart from checkpoint simulation, forget to delete your ICs

## Impact
Wasted time on code coupling work

### PR DESCRIPTION
**Pull request by GiudGiud**
Right now the system lets you specify both. And the IC takes over. On timestep 0, one can see the initial condition appear, overriding the restart, and the system uses the ICs (which changes the result) 

This is unfortunate because users often have ICs in their input because they copied over the steady input to make a transient that restarts from a steady calculation. There generally is not much of a need for ICs in that restarted case unless:

- new fields are introduced and need to be initialized
- a steady solution is not the one we want to start from for the transient

Hence why I added a boolean to work around the error

refs #21423 
Does not close it since no error with just this PR

### CODE REVIEW

**Comment by GiudGiud** [comment]
Applications with tests affected (check when patch is up)
- [ ] BISON
- [ ] Griffin
- [ ] Pronghorn
- [ ] RELAP7
- [ ] Falcon

**Comment by GiudGiud** [comment]
Would love an approval on the design of this before I start making patches for apps

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -2966,6 +2972,22 @@ FEProblemBase::addInitialCondition(const std::string & ic_name,
...
+  if (!_allow_ics_during_restart)
+  {
+    std::string restart_method = "";
+    if (_app.isRestarting())
+      restart_method = "a checkpoint restart";
+    else if (_app.getExReaderForRestart())
+      restart_method = "an Exodus restart";
+    if (!restart_method.empty())
+      mooseError(
+          "Initial condition has been specified during ",
```
Comment (joshuahansel) [inline framework/src/problems/FEProblemBase.C:2985]:
```suggestion
          "Initial conditions have been specified during ",
```

File: test/tests/kernels/ode/coupled_ode_td_auxvar_ic_from_mesh.i [hunk 2]
```diff
@@ -4,68 +4,74 @@
...
     function = function_g
     variable = g
     execute_on = 'timestep_end'
-  [../]
+  []
+[]
+
+[Problem]
+  # There are initial conditions overwriting the restart on the nonlinear variables
+  # However this test is targetted at the auxiliary variable restart so it's ok
```
Comment (joshuahansel) [inline test/tests/kernels/ode/coupled_ode_td_auxvar_ic_from_mesh.i:73]:
```suggestion
  # However this test is targeted at the auxiliary variable restart so it's ok
```

**Review by joshuahansel — CHANGES_REQUESTED** [review]

**Comment by GiudGiud** [comment]
We can for Exodus I think. For checkpoint rn, I dont think so.
I ll add this

File: test/tests/auxkernels/solution_aux/solution_aux_exodus_interp_restart2.i [hunk 3]
```diff
@@ -70,4 +70,7 @@
 
 [Problem]
   restart_file_base = solution_aux_exodus_interp_restart1_out_cp/0005
+  # There are initial conditions overwriting the restart on the nonlinear variables
+  # However this test is targetted at the auxiliary variable initialized from the solution uo so it's ok
```
Comment (GiudGiud) [inline test/tests/auxkernels/solution_aux/solution_aux_exodus_interp_restart2.i:74]:
```suggestion
  # However this test is targeted at the auxiliary variable initialized from the solution uo so it's ok
```

**Review by lindsayad — COMMENTED** [review]
This looks sensical to me

**Review by loganharbour — DISMISSED** [review]
Would be better to make the default true, fix the apps without breaking them, and then change it to false

**Comment by GiudGiud** [comment]
That s a great idea. I never seem to be able to think of these things

**Comment by GiudGiud** [comment]
This should pass now. 

I agree with Josh that it s nice to be more restrictive and when we trigger this error. However:
- with checkpoint restart we are not selecting which variable to restarts, it should be all of them. So adding an initial condition is only a valid thing for a "NEW" variable, which is tough to know which one is new right now (in the future with flexible restarts I suppose it will be easier)

- with exodus restart, as soon as the variable is in the file we have the error. However some people restart only a few of the variables in the file. This is not in the spirit of restart but is very common (see flurry of failures in the apps). I think making people restart more things is good. so having the error in that situation is still a good thing

**Comment by GiudGiud** [comment]
All patches are made. This needs to be merged before any patch can be merged.

**Comment by joshuahansel** [comment]
> This means that if the spelling of a variable is different between exodus and the input file, we wont get the error. So not sure this is better but it s ok with me

I didn't know a variable could have a different spelling. How does this work - how do they get linked?

My suggestion was not necessarily to change the error behavior, but just to give additional useful information when possible. For instance, if people just blindly started copying and pasting `allow_initial_conditions_with_restart = true`. I'm ok with either error behavior, so you choose what's best.

**Review by joshuahansel — APPROVED** [review]

**Comment by GiudGiud** [comment]
> I didn't know a variable could have a different spelling. How does this work - how do they get linked?

I think they dont right now but it s a very artificial limitation

**Pull request merged**